src/plugins/data/server/search/aggs/aggs_service.ts

```diff
@@ -0,0 +1,127 @@
+   * plugins in `setup`. Its contents are merged with the default agg types
+   * and added to the "real" registry which is instantiated when someone calls
+   * `data.search.aggs.asScopedToClient`.
+   */
+  private readonly aggTypesCache = new AggTypesRegistry();
+
+  /**
+   * getForceNow uses window.location on the client, so we must have a
+   * separate implementation of calculateBounds on the server.
+   */
+  private calculateBounds = (timeRange: TimeRange) => calculateBounds(timeRange, {});
+
+  public setup({ registerFunction }: AggsSetupDependencies): AggsSetup {
+    this.registerFunction = registerFunction;
+
+    return {
+      types: this.aggTypesCache.setup(),
+    };
+  }
+
+  public start({ fieldFormats, uiSettings }: AggsStartDependencies): AggsStart {
+    return {
+      asScopedToClient: async (savedObjectsClient: SavedObjectsClientContract) => {
+        const uiSettingsClient = uiSettings.asScopedToClient(savedObjectsClient);
+        const formats = await fieldFormats.fieldFormatServiceFactory(uiSettingsClient);
+
+        // cache ui settings, only including items which are explicitly needed by aggs
+        const uiSettingsCache = pick(await uiSettingsClient.getAll(), aggsRequiredUiSettings);
+        const getConfig = <T = any>(key: string): T => {
+          return uiSettingsCache[key];
+        };
+
+        const defaultAggTypes = getAggTypes({
+          calculateBounds: this.calculateBounds,
+          getConfig,
+          getFieldFormatsStart: () => ({
+            deserialize: formats.deserialize,
+            getDefaultInstance: formats.getDefaultInstance,
+          }),
+          /**
+           * Date histogram and date range need to know whether we are using the
+           * default timezone, but `isDefault` is not currently offered on the
+           * server, so we need to manually check for the default value.
+           */
+          isDefaultTimezone: () => getConfig('dateFormat:tz') === 'Browser',
+        });
+
+        /**
+         * Instantiating a new `AggsCommonService` will create a new registry, which
+         * is desirable in this case as the scoped ui settings need to be provided
+         * to some of the registered default aggs. So we merge the cached agg types
+         * from `setup` with the default aggs that use scoped settings, ensuring
+         * downstream consumers are receiving everything that has been registered.
+         */
+        const aggsCommonService = new AggsCommonService();
+        const cachedAggTypes = this.aggTypesCache.start().getAll();
+
+        // call setup to ensure all default agg types & functions are registered
+        aggsCommonService.setup({
+          aggTypes: {
+            buckets: defaultAggTypes.buckets.concat(cachedAggTypes.buckets),
+            metrics: defaultAggTypes.metrics.concat(cachedAggTypes.metrics),
+          },
+          registerFunction: this.registerFunction!,
```

Comment:
we shouldn't be passing this one in (or even using it anywhere it the start lifecycle). all the functions should be registered in setup lifecycle.

Reply:
seems aggsCommonSetup should be called from a setup method here

Reply:
Summary of the plan we worked out together offline:
- I will move expression functions to register during `setup`. This is a bug and needs to be fixed anyway so that others can rely on the functions in `start`.
- We will change the way the aggs registry works to register the _providers_ for the agg types instead of registering the instantiated types themselves. Then in `start` when looking up an item in the registry, that item will be re-instantiated each time with the dependencies injected. (This pattern only works for registry items that don't store global state, which is already the case for aggs).
  - This pattern will enable us to still register everything during the `setup` lifecycle method, and will also address the issue [I outlined in my comment above](https://github.com/elastic/kibana/pull/74472/files/2b14394f6dd0bffae73e03392c41aec0550eee5d#r466143644) about 3rd party plugins registering items which depend on uiSettings.
  - Eventually we will look to use this pattern elsewhere; in particular it can be used for field formats registry.